packages/next/client/components/layout-router.tsx

```diff
@@ -120,20 +128,21 @@ class ScrollAndFocusHandler extends React.Component<{
     if (focusAndScrollRef.apply && domNode instanceof HTMLElement) {
       // State is mutated to ensure that the focus and scroll is applied only once.
       focusAndScrollRef.apply = false
+
+      // Try scrolling go the top of the document to be backward compatible with pages
+      if (!topLeftOfElementInViewport(domNode)) {
+        handleSmoothScroll(() => window.scrollTo(0, 0))
+      }
+
+      // Scroll to domNode if domNode is not in viewport when scrolled to top of document
+      if (!topLeftOfElementInViewport(domNode)) {
```

Comment:
The two conditions seem to be the same so both will be called 🤔

Reply:
The first condition should move the dom into the viewport. The second condition won't fire then.

Tests are passing so it should work.

Reply:
We could try to guess if we need to navigate but we can cover just the top 95% of cases.
CSS can be wild.

Reply:
Asked @sebmarkbage to have a look.

Reply:
It’s unfortunate that it would scroll somewhere even if we don’t have to. Like a few pixels sometimes to reveal something that’s already there.

One thing we could do is use an intersectionobserver to observe the element and see if it’s sufficiently in the viewport and it not, then scrollIntoView.

The extra gap would also ensure that we wait for all other reflow before scrolling.

Reply:
Oh, completely forgot about `IntersectionObserver` that will even work with nested scrollable containers. Thanks for pointing that out.

But I'm a little confused by
> It’s unfortunate that it would scroll somewhere even if we don’t have to.

Because the current implementation shouldn't scroll if it's not needed. It will scroll only if the top of the page is not visible. The only controversial part is that it might scroll to the top and back (instantly) but it should have always the correct end result.

>The extra gap would also ensure that we wait for all other reflow before scrolling.

I'm not sure if I get this. Are you talking about the time gap when we are waiting for information from the `IntersectionObserver`?

Reply:
`IntersectionObserver` is too slow. It takes more than one frame so it will flicker. (It will show one frame unscrolled, one frame scrolled to the top and the next frame will be correct):
![image](https://user-images.githubusercontent.com/12820484/208431800-c93acc09-0656-4692-91ed-3e0f15dac245.png)
So it flickers even without scrolling to the top.

Let me check the previous implementation but I think that it was not an issue before.

Reply:
The old implementation is synchronous and does all the scrolling inside `componentDidMount`. It will do everything in just 1 ms (it's a very simple example but it's way faster than IntersectionObserver).

![image](https://user-images.githubusercontent.com/12820484/208433482-768c3713-b59b-4e0b-ba94-0944ebd1fa16.png)

Reply:
I haven't read the whole thread but it seems that `IntersectionObserver` is designed in a way that makes it impossible for us to use it synchronously.

If I understood correctly this thread: https://github.com/w3c/IntersectionObserver/issues/133
Then It will fire after the new content is on the screen which is exactly what we don't want.

Reply:
Synchronous != Fast. Synchronous means that if you do more than one thing, you do it more than once. For example, in this case cause reflows more than once. So synchronous means slower. (Same thing in React. Transitions can be faster than synchronous.)

IntersectionObserver should fire before the next rAF so it should be async but before painting. If that's not happening it might be because the React render itself is too late but that since odd since React itself doesn't render a transition during these events. So seems like something is off.

Reply:
:+1: Sorry I worded that poorly.
I know nothing about the internals of `IntersectionObserver` but I understood that it will generate new entries after each rAF: https://github.com/w3c/IntersectionObserver/issues/133#issuecomment-221690658

It shouldn't be a problem caused by react being late since I create a new `IntersectionObserver` for each scroll and even the second one is slow. Ideally, I want just one intersection observer, but it still shouldn't wait for the frame to be rendered each time.

I'll quickly check the specification at https://w3c.github.io/IntersectionObserver

Reply:
When playing around with the `takeRecords` It appears that it will fire after the next rAF:

https://github.com/JanKaifer/next.js/blob/13654cbda9076d7f182ced66c3f12f1aad3efe1b/packages/next/client/components/layout-router.tsx#L107-L128

It will always return those records in that `setTimeout`.
When benchmarked, it will also show one frame before each scroll.

Another weird thing is that for some reason those frames take 30ms to render. But there GPU and main thread are both empty.
![image](https://user-images.githubusercontent.com/12820484/208472777-b5fb537e-c89f-45c4-88c0-7f99952af246.png)

This is the profile of the interaction:
https://gist.github.com/JanKaifer/5916bff7b3a4e88001d2c30e1bee03f9

Reply:
I tried digging into this in Firefox and I wasn't able to trace my code being invoked. But each `IntersectionObserver` query takes also ~30ms to be resolved.

Reply:
Can you share the code of your experiments? That's not what I'm seeing in my simple examples. It just scrolls into view before painting.

Reply:
The rAF callback isn't what's interesting. It's whether or not the browser paints a new screen before scrolling. rAFs can also be tricky because they might not start immediately.

You also shouldn't be using promises or async/await for this, those just mess with the timing of things and get you put on the wrong queues in sequencing.

Reply:
I'd expect the code to look something like:

```
const scrollIntoViewObserver = new IntersectionObserver((entries) => {
  entries.forEach(entry => {
     scrollIntoViewObserver.unobserve(entry.target);
     if (isTopNotInView(entry)) {
       scrollToTop();
       entry.target.scrollIntoView();
     }
  });
}

...
componentDidMount() {
  scrollIntoViewObserver.observe(node);
}
```

Reply:
> You also shouldn't be using promises or async/await for this, those just mess with the timing of things and get you to put on the wrong queues in sequencing.

This is something I didn't realize. I'll try handling promises manually and won't use `async/await`.

This is what I used to test intersection observer. https://github.com/JanKaifer/next.js/blob/13654cbda9076d7f182ced66c3f12f1aad3efe1b/packages/next/client/components/layout-router.tsx#L106-L170
When viewing this commit I manually executed this test:
https://github.com/JanKaifer/next.js/blob/f93f1b1033abd6f0ab93ff81b5db8b64667ceb4d/test/e2e/app-dir/router-autoscroll.test.ts#L66-L78

Reply:
I redid it without `aynsc/await`. It still flashes but it's better because it's just `16ms` per frame now (down from `32ms`).
https://github.com/JanKaifer/next.js/blob/1944a95f7d8e52c307b3c5e4022472b9e329b4f3/packages/next/client/components/layout-router.tsx#L106-L168

Reply:
The callback itself fires always just after (`2ms`) a new frame.
![image](https://user-images.githubusercontent.com/12820484/209095266-60228235-4afa-4d6c-80f3-3e88563f06e3.png)

Reply:
> It's whether or not the browser paints a new screen before scrolling.

I think this is not happening. From the devtools the paint happens before we even start scrolling.

It seems that issue is that the callback from `IntersectionObserver` is called too late.

I will try to use that hack with `takeRecords` and `rAF`. Maybe I'll get it to give us the info sooner (In the previous example it always fired sooner).

Reply:
Firing `takeRecords` in `rAF(setTimeout)` is fired after painting.

Reply:
I'm confused about why you're calling `requestAnimationFrame` at all here: https://github.com/JanKaifer/next.js/blob/13654cbda9076d7f182ced66c3f12f1aad3efe1b/packages/next/client/components/layout-router.tsx#L118

Intersection observers fire toward the end of the current frame and scheduling a rAF from within those would then end up in the next frame which is too late.

So seems you should just not call rAF and use the entries you get int he Intersection observer callback.

Reply:
Btw, none of this should be done when we're navigating back/forward within a popstate, because there we should be able to rely on native scroll restoration (but we probably can't atm because React has a bug that it doesn't flush early enough).

Reply:
> I'm confused about why you're calling requestAnimationFrame at all here: [JanKaifer/next.js@13654cb/packages/next/client/components/layout-router.tsx#L118](https://github.com/JanKaifer/next.js/blob/13654cbda9076d7f182ced66c3f12f1aad3efe1b/packages/next/client/components/layout-router.tsx#L118)

That's what was suggested in that issue I mentioned earlier as the fastest way to get those entries. It behaves the same as a normal callback in `IntersectionObserver`.

Reply:
Maybe if you could look at this one https://github.com/JanKaifer/next.js/blob/1944a95f7d8e52c307b3c5e4022472b9e329b4f3/packages/next/client/components/layout-router.tsx#L106-L168
It uses the `IntersectionObserver` callback and It also doesn't use `async/await`.

Reply:
> Btw, none of this should be done when we're navigating back/forward within a popstate

I think I saw a thread about that we are implementing this scroll restoration with custom code. Anyway good thing to have tests for, will create an issue to write/check them.

Reply:
Let's merge this. We can always migrate to `IntersectionObserver` in a followup PR.

Reply:
> Btw, none of this should be done when we're navigating back/forward within a popstate, because there we should be able to rely on native scroll restoration (but we probably can't atm because React has a bug that it doesn't flush early enough).

oh im so glad to read your comment - do you think anyone of you could help me to finalise this PR? https://github.com/vercel/next.js/pull/45101 I'm pretty sure it touches the right parts.